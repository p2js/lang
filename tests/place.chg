fn place<T>(x: T*): T* { return x; }

fn main(): i32 {
    let x = .{ a: 1 };
    x.a = x.a + 1;

    // op-assign operators with side-effect places
    let v = 2;
    *place(&v) = 4;
    *place(&v) += 4;
    *place(&v) -= 4;
    *place(&v) *= 4;
    *place(&v) /= 4;
    *place(&v) %= 4;

    return x.a;
}
