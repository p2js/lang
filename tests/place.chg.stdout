expr @ 1:33 1:34 => T*
expr @ 1:26 1:34 => never
expr @ 1:24 2:0 => ()
expr @ 4:19 4:20 => i32
expr @ 4:13 4:22 => {a: i32}
expr @ 5:11 5:12 => {a: i32}
expr @ 5:11 5:14 => i32
expr @ 5:17 5:18 => i32
expr @ 5:11 5:18 => i32
expr @ 5:5 5:6 => {a: i32}
expr @ 5:5 5:8 => i32
expr @ 5:5 5:18 => ()
expr @ 8:13 8:14 => i32
expr @ 9:18 9:19 => i32
expr @ 9:6 9:11 => fn place(...)
expr @ 9:13 9:14 => i32
expr @ 9:12 9:14 => i32*
expr @ 9:6 9:15 => i32*
expr @ 9:5 9:15 => i32
expr @ 9:5 9:19 => ()
expr @ 10:19 10:20 => i32
expr @ 10:6 10:11 => fn place(...)
expr @ 10:13 10:14 => i32
expr @ 10:12 10:14 => i32*
expr @ 10:6 10:15 => i32*
expr @ 10:5 10:15 => i32
expr @ 10:5 10:20 => ()
expr @ 11:19 11:20 => i32
expr @ 11:6 11:11 => fn place(...)
expr @ 11:13 11:14 => i32
expr @ 11:12 11:14 => i32*
expr @ 11:6 11:15 => i32*
expr @ 11:5 11:15 => i32
expr @ 11:5 11:20 => ()
expr @ 12:19 12:20 => i32
expr @ 12:6 12:11 => fn place(...)
expr @ 12:13 12:14 => i32
expr @ 12:12 12:14 => i32*
expr @ 12:6 12:15 => i32*
expr @ 12:5 12:15 => i32
expr @ 12:5 12:20 => ()
expr @ 13:19 13:20 => i32
expr @ 13:6 13:11 => fn place(...)
expr @ 13:13 13:14 => i32
expr @ 13:12 13:14 => i32*
expr @ 13:6 13:15 => i32*
expr @ 13:5 13:15 => i32
expr @ 13:5 13:20 => ()
expr @ 14:19 14:20 => i32
expr @ 14:6 14:11 => fn place(...)
expr @ 14:13 14:14 => i32
expr @ 14:12 14:14 => i32*
expr @ 14:6 14:15 => i32*
expr @ 14:5 14:15 => i32
expr @ 14:5 14:20 => ()
expr @ 16:12 16:13 => {a: i32}
expr @ 16:12 16:15 => i32
expr @ 16:5 16:15 => never
expr @ 3:16 18:1 => ()



define i32* @place$LT$i32$GT$(i32* %lt.1) {
prologue:
%l.0 = alloca i32*
%l.1 = alloca i32*
store i32* %lt.1, i32** %l.1
br label %bb.0

bb.0:
%l.2 = load i32*, i32** %l.1
store i32* %l.2, i32** %l.0
%ret.0 = load i32*, i32** %l.0
ret i32* %ret.0

}

define i32 @main() {
prologue:
%l.0 = alloca i32
%l.1 = alloca {i32}
%l.4 = alloca i32
br label %bb.0

bb.0:
%t.0 = alloca {i32}
%p.1 = getelementptr {i32}, {i32}* %t.0, i32 0, i32 0
store i32 1, i32* %p.1
%t.2 = load {i32}, {i32}* %t.0
store {i32} %t.2, {i32}* %l.1
%t.3 = getelementptr {i32}, {i32}* %l.1, i32 0, i32 0
%l.2 = load i32, i32* %t.3
%l.3 = add i32 %l.2, 1
%t.4 = getelementptr {i32}, {i32}* %l.1, i32 0, i32 0
store i32 %l.3, i32* %t.4
store i32 2, i32* %l.4
%l.6 = bitcast i32* %l.4 to i32*
%l.5 = call i32* @place$LT$i32$GT$(i32* %l.6)
br label %bb.1

bb.1:
%t.5 = alloca i32*
store i32* %l.5, i32** %t.5
%t.6 = load i32*, i32** %t.5
store i32 4, i32* %t.6
%l.8 = bitcast i32* %l.4 to i32*
%l.7 = call i32* @place$LT$i32$GT$(i32* %l.8)
br label %bb.2

bb.2:
%t.7 = alloca i32*
store i32* %l.7, i32** %t.7
%t.8 = load i32*, i32** %t.7
%l.9 = load i32, i32* %t.8
%l.10 = add i32 %l.9, 4
%t.9 = alloca i32*
store i32* %l.7, i32** %t.9
%t.10 = load i32*, i32** %t.9
store i32 %l.10, i32* %t.10
%l.12 = bitcast i32* %l.4 to i32*
%l.11 = call i32* @place$LT$i32$GT$(i32* %l.12)
br label %bb.3

bb.3:
%t.11 = alloca i32*
store i32* %l.11, i32** %t.11
%t.12 = load i32*, i32** %t.11
%l.13 = load i32, i32* %t.12
%l.14 = sub i32 %l.13, 4
%t.13 = alloca i32*
store i32* %l.11, i32** %t.13
%t.14 = load i32*, i32** %t.13
store i32 %l.14, i32* %t.14
%l.16 = bitcast i32* %l.4 to i32*
%l.15 = call i32* @place$LT$i32$GT$(i32* %l.16)
br label %bb.4

bb.4:
%t.15 = alloca i32*
store i32* %l.15, i32** %t.15
%t.16 = load i32*, i32** %t.15
%l.17 = load i32, i32* %t.16
%l.18 = mul i32 %l.17, 4
%t.17 = alloca i32*
store i32* %l.15, i32** %t.17
%t.18 = load i32*, i32** %t.17
store i32 %l.18, i32* %t.18
%l.20 = bitcast i32* %l.4 to i32*
%l.19 = call i32* @place$LT$i32$GT$(i32* %l.20)
br label %bb.5

bb.5:
%t.19 = alloca i32*
store i32* %l.19, i32** %t.19
%t.20 = load i32*, i32** %t.19
%l.21 = load i32, i32* %t.20
%l.22 = sdiv i32 %l.21, 4
%t.21 = alloca i32*
store i32* %l.19, i32** %t.21
%t.22 = load i32*, i32** %t.21
store i32 %l.22, i32* %t.22
%l.24 = bitcast i32* %l.4 to i32*
%l.23 = call i32* @place$LT$i32$GT$(i32* %l.24)
br label %bb.6

bb.6:
%t.23 = alloca i32*
store i32* %l.23, i32** %t.23
%t.24 = load i32*, i32** %t.23
%l.25 = load i32, i32* %t.24
%l.26 = srem i32 %l.25, 4
%t.25 = alloca i32*
store i32* %l.23, i32** %t.25
%t.26 = load i32*, i32** %t.25
store i32 %l.26, i32* %t.26
%t.27 = getelementptr {i32}, {i32}* %l.1, i32 0, i32 0
%l.27 = load i32, i32* %t.27
store i32 %l.27, i32* %l.0
%ret.28 = load i32, i32* %l.0
ret i32 %ret.28

}


